internal/machinesavailability/handler.go

```diff
@@ -0,0 +1,183 @@
+		rulesService:   rulesService,
+		gardenerClient: gardenerClient,
+		clientFactory:  clientFactory,
+		logger:         logger.With("service", "MachinesAvailabilityHandler"),
+	}
+}
+
+func (h *Handler) AttachRoutes(router *httputil.Router) {
+	router.HandleFunc("/oauth/machines_availability", h.getMachinesAvailability)
+}
+
+func (h *Handler) getMachinesAvailability(w http.ResponseWriter, req *http.Request) {
+	supportedProviders := []runtime.CloudProvider{runtime.AWS}
+	var providersData ProvidersData
+
+	for _, provider := range supportedProviders {
+		providerEntry := Provider{
+			Name:         provider,
+			MachineTypes: []MachineType{},
+		}
+
+		regionSupportingMachine, err := h.providerSpec.RegionSupportingMachine(string(provider))
+		if err != nil {
+			httputil.WriteErrorResponse(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		accessKeyID, secretAccessKey, err := h.clientCredentials(strings.ToLower(string(provider)))
+		if err != nil {
+			httputil.WriteErrorResponse(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		machineTypes := h.providerSpec.MachineTypes(provider)
+		machineFamilies := make(map[string]string)
+		for _, machineType := range machineTypes {
+			var family string
+			if provider == runtime.AWS {
+				parts := strings.SplitN(machineType, ".", 2)
+				family = parts[0]
+			} else {
+				httputil.WriteErrorResponse(w, http.StatusInternalServerError, fmt.Errorf("%s not supported", provider))
+				return
+			}
+			machineFamilies[family] = machineType
+		}
+
+		for machineFamily, machineType := range machineFamilies {
+			machineTypeEntry := MachineType{
+				Name:    machineFamily,
+				Regions: []Region{},
+			}
+
+			regions := regionSupportingMachine.SupportedRegions(machineType)
+			if len(regions) == 0 {
+				regions = h.providerSpec.Regions(provider)
+			}
+
+			for _, region := range regions {
+				client, err := h.clientFactory.New(context.Background(), accessKeyID, secretAccessKey, region)
+				if err != nil {
+					httputil.WriteErrorResponse(w, http.StatusInternalServerError, err)
+					return
+				}
+
+				count, err := client.AvailableZonesCount(context.Background(), machineType)
+				if err != nil {
+					httputil.WriteErrorResponse(w, http.StatusInternalServerError, err)
+					return
+				}
+
+				highAvailability := count >= 3
+				machineTypeEntry.Regions = append(machineTypeEntry.Regions, Region{
+					Name:             region,
+					HighAvailability: highAvailability,
+				})
+			}
+
+			providerEntry.MachineTypes = append(providerEntry.MachineTypes, machineTypeEntry)
+		}
+
+		sort.Slice(providerEntry.MachineTypes, func(i, j int) bool {
+			return providerEntry.MachineTypes[i].Name < providerEntry.MachineTypes[j].Name
+		})
+
+		providersData.Providers = append(providersData.Providers, providerEntry)
+	}
+
+	httputil.WriteResponse(w, http.StatusOK, providersData)
+}
+
+func (h *Handler) clientCredentials(provider string) (string, string, error) {
```

Comment:
I wonder if dividing this function into three separate part would make sense:
- getting rule
- getting secret binding
- extracting credentials (this is as a matter of fact already done).
Rationale:
 - making it easier to migrate from secret bindings to credential bindings.